dropbear_hook

```diff
@@ -6,7 +6,10 @@ run_hook ()
   mount -t devpts devpts /dev/pts
 
   echo "Starting dropbear"
-  /usr/sbin/dropbear -E -s -j -k
+
+  # Load configuration options (currently only default listen address)
+  [ -r /etc/dropbear/dropbear.conf ] && . /etc/dropbear/dropbear.conf
```

Comment:
I like this usage, but it is also pretty brittle. It seems like it would be really easy to have a syntax problem in the conf file that results in any number of dead-ends that make the system unbootable. It seems like something should be done at the time the file is copied (during settup instead of boot) to verify the file syntax, at least so that it doesn't throw a shell parsing error or hang.

Reply:
There isn't really a good way to verify syntax of the file without simply sourcing the file with `${BUILDROOT}/bin/sh` in the `build` script. However, this is a bad idea because some legitimate uses of this file may produce side effects that should not be triggered in the building environment.

Most syntax errors will simply render the file unable to be sourced, which will leave `$dropbear_listen` unset even if the file otherwise attempts to set it. The biggest risk is some user inadvertently adding `exit` or an `exec` to the file, which will take down `init` and the kernel will panic.

I am open to the argument that the source and dropbear invocation should happen in a subshell, which will guard against `exec` and `exit` mishaps. How many pads do we offer a user (who, to be using this hook, is probably reasonably advanced) as a defense against a footgun?

Reply:
Fair enough, I see your point(s). What would a subshell wrapper look like? It might be than any level of footgun protection we provide would just add complexity and confusion and letting the most obvious footgun live on is the best solution.

Reply:
Wrapping this in a subshell is just putting parentheses around the source and dropbear execution lines:
```bash
(
  [ -r /etc/dropbear/dropbear.conf ] && . /etc/dropbear/dropbear.conf
  /usr/sbin/dropbear -E -s -j -k -p "${dropbear_listen:-22}"
)
```

Reply:
Oh ... right ... I was trying to figure out how a subshell would pass variables out of the subshell. Putting the dropbeard in a subshell sidesteps that conundrum.

I would say putting the whole thing in in a subshell like that might be worthwhile ... depending on how the hook itself is processed. If something in the hook here goes sideways, does it block the rest of the boot process (and e.g. console unlock ability)? If so then I think that level of protection is probably a good idea. If a bad exit at that point won't hold anything else anyway then what's the point?

Reply:
There are no real negative effects of running these commands in a subshell. It's possible somebody could put something like, *e.g.*, `sleep 9999999` in `dropbear.conf` and prevent the hook from returning, which will stall everything else that follows it. Of course, that would happen even without the subshell.

The biggest benefit to running in a subshell is that people can use (abuse) `dropbear.conf` to prevent `dropbear` from running or short-circuit its configuration based on run-time parameters. Consider, for example, the following `dropbear.conf`:
```bash
if ip link show dev eth1 >/dev/null 2>&1; then
    return;
elif ! ip link show dev eth0 >/dev/null 2>&1; then
    exit;
fi
dropbear_listen=2222
```
This is a totally contrived example that will cause dropbear to listen to the default port 22 if `eth1` is found (aborting all further processing of the configuration) or, if not, completely prevent dropbear from running unless `eth0` is present. If `eth0` is present (but not `eth1`), dropbear will listen on port 2222.